features/mbedtls/targets/TARGET_CRYPTOCELL310/aes_alt.c

```diff
@@ -0,0 +1,1061 @@
+
+                RK[7]  = RK[1] ^ RK[6];
+                RK[8]  = RK[2] ^ RK[7];
+                RK[9]  = RK[3] ^ RK[8];
+                RK[10] = RK[4] ^ RK[9];
+                RK[11] = RK[5] ^ RK[10];
+            }
+            break;
+
+        case 14:
+
+            for( i = 0; i < 7; i++, RK += 8 )
+            {
+                RK[8]  = RK[0] ^ RCON[i] ^
+                ( (uint32_t) FSb[ ( RK[7] >>  8 ) & 0xFF ]       ) ^
+                ( (uint32_t) FSb[ ( RK[7] >> 16 ) & 0xFF ] <<  8 ) ^
+                ( (uint32_t) FSb[ ( RK[7] >> 24 ) & 0xFF ] << 16 ) ^
+                ( (uint32_t) FSb[ ( RK[7]       ) & 0xFF ] << 24 );
+
+                RK[9]  = RK[1] ^ RK[8];
+                RK[10] = RK[2] ^ RK[9];
+                RK[11] = RK[3] ^ RK[10];
+
+                RK[12] = RK[4] ^
+                ( (uint32_t) FSb[ ( RK[11]       ) & 0xFF ]       ) ^
+                ( (uint32_t) FSb[ ( RK[11] >>  8 ) & 0xFF ] <<  8 ) ^
+                ( (uint32_t) FSb[ ( RK[11] >> 16 ) & 0xFF ] << 16 ) ^
+                ( (uint32_t) FSb[ ( RK[11] >> 24 ) & 0xFF ] << 24 );
+
+                RK[13] = RK[5] ^ RK[12];
+                RK[14] = RK[6] ^ RK[13];
+                RK[15] = RK[7] ^ RK[14];
+            }
+            break;
+    }
+    ctx->sw_key_set = 1;
+
+    return( 0 );
+
+}
+
+//taken from the aes.c function mbedtls_aes_setkey_dec
+int mbedtls_aes_setkey_sw_dec( struct sw_aes_context *ctx, const unsigned char *key,
+                    unsigned int keybits )
+{
+    int i, j, ret;
+    struct sw_aes_context cty;
+    uint32_t *RK;
+    uint32_t *SK;
+
+    memset ( &cty, 0, sizeof( struct sw_aes_context ) );
+
+    ctx->rk = RK = ctx->buf;
+
+    /* Also checks keybits */
+    if( ( ret = mbedtls_aes_setkey_sw_enc( &cty, key, keybits ) ) != 0 )
+        goto exit;
+
+    ctx->nr = cty.nr;
+
+
+    SK = cty.rk + cty.nr * 4;
+
+    *RK++ = *SK++;
+    *RK++ = *SK++;
+    *RK++ = *SK++;
+    *RK++ = *SK++;
+
+    for( i = ctx->nr - 1, SK -= 8; i > 0; i--, SK -= 8 )
+    {
+        for( j = 0; j < 4; j++, SK++ )
+        {
+            *RK++ = RT0[ FSb[ ( *SK       ) & 0xFF ] ] ^
+                    RT1[ FSb[ ( *SK >>  8 ) & 0xFF ] ] ^
+                    RT2[ FSb[ ( *SK >> 16 ) & 0xFF ] ] ^
+                    RT3[ FSb[ ( *SK >> 24 ) & 0xFF ] ];
+        }
+    }
+
+    *RK++ = *SK++;
+    *RK++ = *SK++;
+    *RK++ = *SK++;
+    *RK++ = *SK++;
+    ctx->sw_key_set = 1;
+
+exit:
+    memset ( &cty, 0, sizeof( struct sw_aes_context ) );
+    return( ret );
+}
+
+int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx, const unsigned char *key,
+                    unsigned int keybits )
+{
+    if (ctx == NULL )
+        return -1;
+
+    if (keybits/8 > SASI_AES_KEY_MAX_SIZE_IN_BYTES)
+        return mbedtls_aes_setkey_sw_enc(&ctx->sw_context,key,keybits);
+
+    ctx->cc_aes_context.CC_cipherFlag = SASI_AES_ENCRYPT;
+    ctx->cc_aes_context.CC_KeyData.pKey = ( unsigned char* ) key;
```

Comment:
Ok, thanks. As you say, this might be a problem. I suggest running the following tests:
1. Encrypt the data using the acclerator
1. Decrypt the output from the previous step using the software implementation from the library.
1. Check that the result matches the input.
Another test is to do the inverse i.e. encrypt using software implementation and decrypt using the accelerator.